### PR TITLE
Add fps calculator for VP8 and DependencyDescriptor

### PR DESCRIPTION
**Pull request by cnderrauber**
Temporal layer subscribe setting based on fps:
1. fps calculator for VP8 and DependencyDescriptor, at the beginning 1~2 second of publish
2. subscribe request video stream that match fps.

The calculator only work at the beginning of publish since dynacast and network jitter would obstruct the process.
If network is good, fps calculator can get results near the client's maxframerate of encoding parameter, but in case of bad network, it will get a lower value that cause client get higher temporal layer than expected. Not a critical issue but it is a point need improve.

### CODE REVIEW

File: pkg/sfu/receiver.go [hunk 1]
```diff
@@ -102,9 +104,6 @@ type WebRTCReceiver struct {
 	// update stats
 	onStatsUpdate func(w *WebRTCReceiver, stat *livekit.AnalyticsStat)
 
-	// update layer info
-	onMaxLayerChange func(maxLayer int32)
```
Comment (cnderrauber) [inline pkg/sfu/receiver.go:106]: not used

File: pkg/sfu/streamtracker.go [hunk 2]
```diff
@@ -184,7 +184,7 @@ func (s *StreamTracker) SetPaused(paused bool) {
 }
 
 // Observe a packet that's received
-func (s *StreamTracker) Observe(sn uint16, temporalLayer int32, pktSize int, payloadSize int) {
```
Comment (cnderrauber) [inline pkg/sfu/streamtracker.go:187]: sn is not used

File: pkg/sfu/receiver.go [hunk 3]
```diff
@@ -613,3 +611,10 @@ func (w *WebRTCReceiver) GetPrimaryReceiverForRed() TrackReceiver {
 	}
 	return w.primaryReceiver.Load().(*RedPrimaryReceiver)
 }
+
+func (w *WebRTCReceiver) GetTemporalLayerFpsForSpatial(layer int32) []float32 {
+	if !w.isSVC {
```
Comment (boks1971) [inline pkg/sfu/receiver.go:616]: Should this be `if w.isSVC`?

**Review by boks1971 — COMMENTED** [review]
@cnderrauber I have not looked at FPS calculator code. Wondering if it can be simplified to just look at temporal layer id and RTP frame marker (at least for VP8). Does that not work at all? I have done that before and it was quite a decent measurement for frame rate. Just wanted to understand before looking at fps.go and fps_test.go.

**Comment by boks1971** [comment]
Another thought. If we can measure it using temporal layer id and RTP frame marker, we can continuously measure it just like the stream tracker measures bitrates now. With stream tracker and dynacast, unneeded spatial layers will be turned off anyway. We can use fps in active layers to adapt temporally. That will work if fps changes over time also. It can also work well if we use RTP frame marking header extension for E2EE also down the line.

**Comment by cnderrauber** [comment]
> @cnderrauber I have not looked at FPS calculator code. Wondering if it can be simplified to just look at temporal layer id and RTP frame marker (at least for VP8). Does that not work at all? I have done that before and it was quite a decent measurement for frame rate. Just wanted to understand before looking at fps.go and fps_test.go.

If we only count on frame marker and temporal id, we got fps is the server side received rates that can be changed by network (lost, jitter, resent). That will have a difference with the publisher send frame rates. I do the calculation in this PR is want fps like resolution (w&h) is a fixed value in the publisher side encodings, we calculate max spatial / temporal layers from that and subscribe setting . The max spatial/temporal is fixed by subscriber request, and the actually send layer is limit by both network condition and max layer. If the fps rate is not fixed and not same as the publisher side rates, then temporal layer will have two variables condition to change, is not same as spatial layer.

File: pkg/sfu/receiver.go (hunk 3 above)
Comment (cnderrauber) [reply pkg/sfu/receiver.go:616]: In non svc encoding, every spatial layer have a standalone buffer, in its own buffer, the spatial is treat as 0 by frame calculate (in ExtPacket, spatial is invalid layer ).

Comment (boks1971) [reply pkg/sfu/receiver.go:616]: Got it.

**Comment by boks1971** [comment]
Thank you @cnderrauber . Will have a look at the FPS code when my brain is fully awake :-)

File: pkg/sfu/buffer/fps_test.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+			nextTs[i][j] = startTs
+			tsStep[i][j] = uint32(90000 / fps[i][j])
+		}
+	}
+
+	currentTs := make([]uint32, spatials)
+	for i := 0; i < spatials; i++ {
+		currentTs[i] = startTs
+	}
+	// currentTs := startTs
```
Comment (boks1971) [inline pkg/sfu/buffer/fps_test.go:69]: can remove commented out stuff.

**Review by boks1971 — COMMENTED** [review]
Don't understand the DD stuff well as I do not know all the details of DD. I need to learn that. A few small comments on the rest.

File: pkg/sfu/buffer/fps.go [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+	if f.baseFrame == nil {
+		f.baseFrame = &frameInfo{seq: ep.Packet.SequenceNumber, ts: ep.Packet.Timestamp, fn: fn}
+		f.fnReceived[0] = f.baseFrame
+		f.firstFrames[spatial][temporal] = f.baseFrame
+		f.secondFrames[spatial][temporal] = f.baseFrame
+		return false
+	}
+
+	baseDiff := fn - f.baseFrame.fn
+	if baseDiff == 0 || baseDiff > 0x8000 {
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply pkg/sfu/buffer/fps.go:262]: yeah, it's uint16

File: pkg/sfu/buffer/fps.go [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+func (f *FrameRateCalculatorDD) SetMaxLayer(spatial, temporal int32) {
+	f.maxSpatial, f.maxTemporal = spatial, temporal
+}
+
+func (f *FrameRateCalculatorDD) RecvPacket(ep *ExtPacket) bool {
+	if f.completed {
+		return true
+	}
+	if ep.Temporal >= int32(len(f.frameRates)) {
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply pkg/sfu/buffer/fps.go:231]: Good catch, already have a check below, so removed it

File: pkg/sfu/buffer/fps.go [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+
+	if f.baseFrame == nil {
+		f.baseFrame = &frameInfo{seq: ep.Packet.SequenceNumber, ts: ep.Packet.Timestamp, fn: fn}
+		f.fnReceived[0] = f.baseFrame
+		f.firstFrames[temporal] = f.baseFrame
+		return false
+	}
+
+	baseDiff := fn - f.baseFrame.fn
+	if baseDiff == 0 || baseDiff > 0x8000 {
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply pkg/sfu/buffer/fps.go:79]: Will change it , thanks for the information

File: pkg/sfu/buffer/fps.go [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+		return false
+	}
+
+	if baseDiff >= uint16(len(f.fnReceived)) {
+		// frame number is not continuous, reset
```
[the remark this answers is not in the document]
Comment (cnderrauber) [reply pkg/sfu/buffer/fps.go:84]: Not call it here since `close` means terminate but here is a restart. Will rename it to reset and call it here.

**Review by boks1971 — APPROVED** [review]

**Pull request merged**